### ISSUE TITLE
Do something when a root command's flag is set

### ISSUE THREAD

**Issue by westhomas** [DannyBen/bashly#359]
### Description

First, great project. Love it. 👍 

I'm trying to figure out the best way to do something, but I'm having trouble.

I have a `--debug` flag on the root command, with no arguments. When this flag is passed I want to do something (i.e. `set -x`).

I was just going to check if debug was set with `if [[ ${args[--debug]} ]]; then ... `, but I can't seem to do this in `initialize.sh` because `${args[--debug]}` doesn't exist yet and throws the error `args: bad array subscript`.

I've tried doing this in a `validate:` yaml directive, but can't get that working either.

Since the `root_command` option was removed when you have sub-commands, I can't use that now either.

I can't seem to find a good way to run some code when the `--debug` flag is set.

Any suggestions?

**Comment by DannyBen** [comment]
Thank you.

1. Well, when you have subcommands, the `root_command` will never be called, this is why it is not there.
2. As you noticed, the `initialize` is not the place - it is done before anything else.

Defining flags on the root, just makes these flags available in all subcommands. Note however, that input order matters:

```
$ <command> <root flags> <subcommand> <subcommand flags and args>
```

With this bashly.yml

```yaml
name: cli
help: Sample application
version: 0.1.0

flags:
- long: --debug
  short: -d
  help: Enable debug mode

commands:
- name: download
  alias: d
  help: Download a file

  flags:
  - long: --force
    short: -f
    help: Overwrite existing files
```

you get:

```
$ ./cli --debug download --force
# this file is located in 'src/download_command.sh'
# code for 'cli download' goes here
# you can edit it freely and regenerate (it will not be overwritten)
args:
- ${args[--debug]} = 1
- ${args[--force]} = 1
```

Does this help?
Was this not intuitive/clear enough in the docs? Let me know where I can improve the behavior or documentation.

**Comment by DannyBen** [comment]
> but I think it would be nice to have a way to do this by convention.

I thought about this, but decided against.
In Rails, here is a concept of `before` and `after` filter, which is similar to what you are describing - since this is a "before" filter.

I found this convention to be more trouble than its worth, and it "gets in your way" more often than not.

The straight forward solution would be to add a library function (`bashly add lib`), and call it from any command file that needs it.

That said, your use case of using `--debug` to `set -x` indeed shows the weak spot of the decision to avoid some sort of "before" filter.

Let me sleep on it.

**Comment by DannyBen** [comment]
This is now implemented in the [edge](edge#readme) version.

You can run `bashly add hooks` to add the two before/after hook files, and edit them as you see fit.

**Comment by DannyBen** [comment]
@westhomas were you able to test it?

**Comment by DannyBen** [comment]
Thanks for raising this issue Wes.

This feature is now released in 1.0.2

Docs:

- [Writing Your Scripts > Before/after hooks](https://bashly.dannyb.co/usage/writing-your-scripts/#beforeafter-hooks)
- [Advanced Features > Before/After Hooks](https://bashly.dannyb.co/advanced/hooks/)
- [Examples > Before/After Hooks Example](https://github.com/DannyBen/bashly/tree/master/examples/hooks#readme)
